utils/prompt_generation.py

```diff
@@ -0,0 +1,424 @@
+import logging
+import argparse
+import json
+from datetime import date
+
+import torch
+from jinja2 import Template
+from datasets import load_dataset
+from transformers import AutoTokenizer
+
+
+logging.basicConfig(
+    level=logging.INFO, format="%(asctime)s - %(name)s - %(levelname)s - %(message)s"
+)
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+
+# set torch seed for reproducibility
+torch.manual_seed(42)
+
+
+def load_alpaca_eval_dataset_samples(num_prompts):
+    # Load alpaca_eval dataset with specified number of samples
+    alpaca_ds = load_dataset(
+        "tatsu-lab/alpaca_eval",
+        "alpaca_eval",
+        split=f"eval[:{num_prompts}]",
+        trust_remote_code=True,
+    )
+    return alpaca_ds["instruction"]
+
+
+def tokenize_encode(prompt, tokenizer, max_length, tokenizer_model):
+    if tokenizer_model == "meta-llama/Llama-3.1-70B-Instruct":
+        return tokenizer.encode(
+            prompt, add_special_tokens=False, truncation=True, max_length=max_length
+        )
+    else:
+        raise ValueError(f"Unsupported tokenizer model: '{tokenizer_model}'.")
+
+
+def tokenize_decode(encoded_prompt, tokenizer, tokenizer_model):
+    if tokenizer_model == "meta-llama/Llama-3.1-70B-Instruct":
+        return tokenizer.decode(encoded_prompt)
+    else:
+        raise ValueError(f"Unsupported tokenizer model: '{tokenizer_model}'.")
+
+
+# Define a function to generate random prompts using a model's vocabulary
+def generate_random_prompts(
+    num_prompts: int,
+    max_length: int,
+    input_seq_len: int,
+    distribution: str,
+    tokenizer_model=None,
+):
+    """
+    Generate random prompts using the model's vocabulary.
+
+    Args:
+        num_prompts (int): Number of prompts to generate.
+        max_length (int, optional): Maximum length of the generated prompt.
+        input_seq_len (int, optional): Length parameter of the input sequence.
+        distribution (str, optional): The distribution method for selecting random prompt lengths ('fixed', 'uniform', or 'normal).
+        tokenizer_model (str, optional): The model to use for generating random tokens.
+
+    Returns:
+        List[str]: A list of generated prompts.
+    """
+    if tokenizer_model is None:
+        raise ValueError("Model must be provided when using 'random' as the dataset.")
+
+    tokenizer = AutoTokenizer.from_pretrained(tokenizer_model)
+    vocab_size = tokenizer.vocab_size
+
+    # Determine the length of each prompt based on the distribution
+    if distribution == "fixed":
+        prompt_lengths = [input_seq_len] * num_prompts
+    elif distribution == "uniform":
+        prompt_lengths = torch.randint(1, input_seq_len, (num_prompts,)).tolist()
+    elif distribution == "normal":
+        prompt_lengths = (
+            torch.normal(mean=input_seq_len, std=input_seq_len / 4, size=(num_prompts,))
+            .clamp(1, max_length)
+            .round()
+            .to(torch.int32)
```

Comment:
Any reason why casting "normal" to `torch.int32` but not "uniform"?

Reply:
yes, uniform uses `torch.randint` so its already an int. Both need to be ints because these are the prompt context lengths.